src/hal/src/queue/family.rs

```diff
@@ -0,0 +1,106 @@
+//! Queue family and groups.
+
+use Backend;
+use queue::{CommandQueue, QueueType};
+use queue::capability::{Capability, Graphics, Compute};
+
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::marker::PhantomData;
+
+/// General information about a queue family, available upon adapter discovery.
+///
+/// *Note*: A backend can expose multiple queue families with the same properties.
+pub trait QueueFamily: Debug {
+    /// Returns the type of queues.
+    fn queue_type(&self) -> QueueType;
+    /// Returns maximum number of queues created from this family.
+    fn max_queues(&self) -> usize;
+    /// Returns true if the queue supports graphics operations.
+    fn supports_graphics(&self) -> bool {
+        Graphics::supported_by(self.queue_type())
+    }
+    /// Returns true if the queue supports graphics operations.
+    fn supports_compute(&self) -> bool {
+        Compute::supported_by(self.queue_type())
+    }
+}
+
+/// Identifier for a queue family of a physical device.
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub struct QueueFamilyId(pub usize);
+
+// Only needed for backend implementations.
```

Comment:
Perhaps, we could have a special module `hal::backend` where we re-export all the things that are needed for backends only, so that we can see all this semi-legal public API in one place?

Reply:
Sounds reasonable!